src/logger.c

```diff
@@ -40,59 +40,62 @@ static const char *logident = NULL;
 bool logcontrol = false;
 int umbilical = 0;
 
+static bool should_log(int level) {
+	// Bail out early if there is nothing to do.
+
+	if(!logcontrol && (level > debug_level || logmode == LOGMODE_NULL)) {
+		return false;
+	}
+
+	return level <= debug_level;
+}
+
 static void real_logger(int level, int priority, const char *message) {
 	char timestr[32] = "";
 	static bool suppress = false;
 
-	// Bail out early if there is nothing to do.
 	if(suppress) {
 		return;
 	}
 
-	if(!logcontrol && (level > debug_level || logmode == LOGMODE_NULL)) {
-		return;
-	}
-
-	if(level <= debug_level) {
-		switch(logmode) {
-		case LOGMODE_STDERR:
-			fprintf(stderr, "%s\n", message);
-			fflush(stderr);
-			break;
+	switch(logmode) {
```

Comment:
There needs to be an `if (level <= debug_level)` around this switch and the first `if`-statement after it, in order to correctly handle the case where `logcontrol == true`.

Reply:
FYI at no point in the code is `logcontrol` documented. Unless you were a seasoned developer I would expect it to be an unknown code quirk.

Reply:
And is there a reason that REQ_LOG doesnt just change the log level? Slight difference in behaviour but simpler and potentially more powerful.

Reply:
You're right that the function of `logcontrol` is not well documented. There are multiple log outputs possible at the same time, with different log levels. The normal log output (to `stdout`, a file or to the syslog) is controlled by the global variable `debug_level`. However, you can use the `tinc log <level>` command to connect to a running tinc daemon and get a stream of log messages with a different log level. Whenever there is at least one control connection requesting log information, `logcontrol` is `true`. That's when we have to do extra work for every call to `logger()` to find out which connections want which log message.

Reply:
There's also `REQ_SET_DEBUG` which allows you to change the global log level of a running tinc daemon.